internal/gps/pkgtree/digest.go

```diff
@@ -0,0 +1,519 @@
+			shiftCount++
+		}
+		nr -= shiftCount // shorten byte read count by number of shifts executed
+
+		// When final byte from a read operation is CR, do not emit it until
+		// ensure first byte on next read is not LF.
+		if f.prevReadEndedCR = buf[nr-1] == '\r'; f.prevReadEndedCR {
+			nr-- // pretend byte was never read from source
+		}
+	} else if f.prevReadEndedCR {
+		// Reading from source returned nothing, but this struct is sitting on a
+		// trailing CR from previous Read, so let's give it to client now.
+		buf[0] = '\r'
+		nr = 1
+		er = nil
+		f.prevReadEndedCR = false // prevent infinite loop
+	}
+	return nr, er
+}
+
+// sortedChildrenFromDirname returns a lexicographically sorted list of child
+// nodes for the specified directory.
+func sortedChildrenFromDirname(osDirname string) ([]string, error) {
+	fh, err := os.Open(osDirname)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot Open")
+	}
+	osChildrenNames, err := sortedChildrenFromDirectorFileHandle(fh)
+	// Close the file handle to the open directory without masking possible
+	// previous error value.
+	if er := fh.Close(); err == nil {
+		err = errors.Wrap(er, "cannot Close")
+	}
+	return osChildrenNames, err
+}
+
+// sortedChildrenFromDirectorFileHandle returns a lexicographically sorted list
+// of child nodes for the specified file handle to a directory.
+func sortedChildrenFromDirectorFileHandle(fh *os.File) ([]string, error) {
+	osChildrenNames, err := fh.Readdirnames(0) // 0: read names of all children
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot Readdirnames")
+	}
+	sort.Strings(osChildrenNames)
+	return osChildrenNames, nil
+}
+
+// writeBytesWithNull appends the specified data to the specified hash, followed by
+// the NULL byte, in order to make accidental hash collisions less likely.
+func writeBytesWithNull(h hash.Hash, data []byte) {
+	// Ignore return values from writing to the hash, because hash write always
+	// returns nil error.
+	_, _ = h.Write(append(data, 0))
+}
+
+// DigestFromDirectory returns a hash of the specified directory contents, which
+// will match the hash computed for any directory on any supported Go platform
+// whose contents exactly match the specified directory.
+//
+// This function ignores any file system node named `vendor`, `.bzr`, `.git`,
+// `.hg`, and `.svn`, as these are typically used as Version Control System
+// (VCS) directories.
+//
+// Other than the `vendor` and VCS directories mentioned above, the calculated
+// hash includes the pathname to every discovered file system node, whether it
+// is an empty directory, a non-empty directory, empty file, non-empty file, or
+// symbolic link. If a symbolic link, the referent name is included. If a
+// non-empty file, the file's contents are included. If a non-empty directory,
+// the contents of the directory are included.
+//
+// While filepath.Walk could have been used, that standard library function
```

Comment:
The above benchmark was performed on one of my systems that has ~1.3 GiB total of file data.

I cannot imagine the performance degradation is due to the additional overhead of calling the anonymous function, which happens to close over a few variables, four of them to be exact.

Because `DirWalk` is really an excellent drop in replacement for `filepath.Walk` for when you need to visit every file system node, it makes it very attractive to keep `DirWalk` around for other portions of this project which may need it.

Just for performance comparison sake, I also benchmarked `DigestFromDirectoryDirWalk` against `DigestFromDirectoryWalk`, which invokes `filepath.Walk`. (I know `filepath.Walk` does not provide the operational functionality we need for this application, however I wanted to determine whether I was doing something completely wrong. Interestingly, `DigestFromDirectoryDirWalk` and `DigestFromDirectoryWalk` performed nearly identically. So perhaps the extra overhead of making the anonymous function call *is* causing the 1.8x slow down for my particular data set.


![screen shot 2017-08-15 at 3 20 42 am](https://user-images.githubusercontent.com/155788/29305851-c910d10e-8168-11e7-8c40-a0c1f41402fb.png)